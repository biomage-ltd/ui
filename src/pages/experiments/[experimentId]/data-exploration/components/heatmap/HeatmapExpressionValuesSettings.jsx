import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Radio } from 'antd';
import PropTypes from 'prop-types';
import updateGeneExpressionType from '../../../../../../redux/actions/genes/updateGeneExpressionType';

const HeatmapExpressionValuesSettings = (props) => {
  const {
    experimentId,
  } = props;

  const dispatch = useDispatch();
  const expressionType = useSelector((state) => state.genes.expression.expressionType);

  const expressionTypes = {
    raw: 'Raw values',
    zScore: 'Z-score',
  };

  const radioStyle = {
    display: 'block',
    height: '30px',
    lineHeight: '30px',
  };

  return (
    <div style={{ padding: '5px' }}>
      <Radio.Group
        value={expressionType}
        onChange={({ target }) => dispatch(updateGeneExpressionType(experimentId, target.value))}
      >
        {Object.keys(expressionTypes).map((type) => (
          <Radio style={radioStyle} value={type}>
            {expressionTypes[type]}
          </Radio>
        ))}
      </Radio.Group>
    </div>
  );
};

HeatmapExpressionValuesSettings.defaultProps = {
};

HeatmapExpressionValuesSettings.propTypes = {
  experimentId: PropTypes.string.isRequired,
};

export default HeatmapExpressionValuesSettings;
