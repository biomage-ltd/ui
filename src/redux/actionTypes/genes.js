const GENES = 'genes';

/**
 * Turns on the loading condition for gene information.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const GENES_PROPERTIES_LOADING = `${GENES}/propertiesLoading`;

/**
 * Sets the state of the gene data to be successfully loaded, with the appropriate gene data.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const GENES_PROPERTIES_LOADED_PAGINATED = `${GENES}/propertiesLoadedPaginated`;

/**
 * Sets an error condition for gene properties data.
 */
const GENES_PROPERTIES_ERROR = `${GENES}/propertiesError`;

/**
 * Updates the list of selected genes with a new selected gene.
 */
const GENES_SELECT = `${GENES}/select`;

/**
 * Removes a deselected gene from the list of selected genes.
 */
const GENES_DESELECT = `${GENES}/deselect`;

/**
 * Turns on the loading state for expression data.
 */
const GENES_EXPRESSION_LOADING = `${GENES}/expressionLoading`;

/**
 * Sets the state of the expression store to be successfully loaded,
 * with the appropriate expression data.
 */
const GENES_EXPRESSION_LOADED = `${GENES}/expressionLoaded`;

/**
 * Sets an error condition for gene expression data.
 */
const GENES_EXPRESSION_ERROR = `${GENES}/expressionError`;

/**
 * Update the selected gene expression type
 */
const GENES_EXPRESSION_TYPE_UPDATE = `${GENES}/expressionTypeUpdate`;

export {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  GENES_EXPRESSION_TYPE_UPDATE,
};
