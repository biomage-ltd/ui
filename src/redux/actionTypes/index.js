const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
