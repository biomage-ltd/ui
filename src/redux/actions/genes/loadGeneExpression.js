import {
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_ERROR, GENES_EXPRESSION_LOADED,
} from '../../actionTypes/genes';

import { fetchCachedWork } from '../../../utils/cacheRequest';

const loadGeneExpression = (
  experimentId, genes, componentUuid, forceReloadAll = false,
) => async (dispatch, getState) => {
  const {
    loading, data: geneData,
  } = getState().genes.expression;

  // If other gene expression data is already being loaded, don't dispatch.
  if (loading.length > 0) {
    return null;
  }

  console.log('genes updating with ', genes);

  // Dispatch loading state.
  dispatch({
    type: GENES_EXPRESSION_LOADING,
    payload: {
      experimentId,
      genes,
      componentUuid,
    },
  });

  // Check which of the genes we actually need to load. Only do this if
  // we are not forced to reload all of the data.
  let genesToFetch = [...genes];

  if (!forceReloadAll) {
    const genesAlreadyLoaded = new Set(Object.keys(geneData));
    genesToFetch = genesToFetch.filter((gene) => !genesAlreadyLoaded.has(gene));
  }

  if (genesToFetch.length === 0) {
    return dispatch({
      type: GENES_EXPRESSION_LOADED,
      payload: {
        experimentId,
        componentUuid,
        genes,
      },
    });
  }

  const body = {
    name: 'GeneExpression',
    genes: genesToFetch,
  };

  try {
    const data = await fetchCachedWork(experimentId, 30, body);
    if (Object.keys(data).length === 0) {
      throw Error('There is no information available for selected genes.');
    }
    dispatch({
      type: GENES_EXPRESSION_LOADED,
      payload: {
        experimentId,
        data,
        componentUuid,
        genes,
      },
    });
  } catch (e) {
    dispatch({
      type: GENES_EXPRESSION_ERROR,
      payload: {
        experimentId,
        error: "Couldn't fetch expression data.",
        componentUuid,
      },
    });
  }
};

export default loadGeneExpression;
