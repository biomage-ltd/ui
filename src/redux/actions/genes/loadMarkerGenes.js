import {
  MARKER_GENES_ERROR, MARKER_GENES_LOADING,
} from '../../actionTypes/genes';
import sendWork from '../../../utils/sendWork';

const REQUEST_TIMEOUT = 60;

const loadMarkerGenes = (experimentId, resolution = null) => async (dispatch, getState) => {
  const {
    backendStatus,
    processing,
  } = getState().experimentSettings;

  let resolutionToSend = resolution;

  if (!resolution) {
    resolutionToSend = processing
      .configureEmbedding.clusteringSettings.methodSettings.louvain.resolution;
  }

  const body = {
    name: 'MarkerHeatmap',
    nGenes: 2,
    type: 'louvain',
    config: {
      resolution: resolutionToSend,
    },
  };

  dispatch({
    type: MARKER_GENES_LOADING,
  });

  try {
    await sendWork(experimentId, REQUEST_TIMEOUT, body, backendStatus.status);
  } catch (e) {
    dispatch({
      type: MARKER_GENES_ERROR,
      payload: {
        experimentId,
        error: e,
      },
    });
  }
};

export default loadMarkerGenes;
