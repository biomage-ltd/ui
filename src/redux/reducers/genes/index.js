import initialState from './initialState';
import {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  GENES_EXPRESSION_TYPE_UPDATE,
} from '../../actionTypes/genes';

import genesPropertiesLoading from './genesPropertiesLoading';
import genesPropertiesLoadedPaginated from './genesPropertiesLoadedPaginated';
import genesPropertiesError from './genesPropertiesError';
import genesExpressionLoading from './genesExpressionLoading';
import genesExpressionLoaded from './genesExpressionLoaded';
import genesExpressionError from './genesExpressionError';
import genesExpressionTypeUpdate from './genesExpressionTypeUpdate';

import genesSelect from './genesSelect';
import genesDeselect from './genesDeselect';

const genesReducer = (state = initialState, action) => {
  switch (action.type) {
    case GENES_PROPERTIES_LOADING: {
      return genesPropertiesLoading(state, action);
    }

    case GENES_PROPERTIES_LOADED_PAGINATED: {
      return genesPropertiesLoadedPaginated(state, action);
    }

    case GENES_PROPERTIES_ERROR: {
      return genesPropertiesError(state, action);
    }

    case GENES_SELECT: {
      return genesSelect(state, action);
    }

    case GENES_DESELECT: {
      return genesDeselect(state, action);
    }

    case GENES_EXPRESSION_LOADING: {
      return genesExpressionLoading(state, action);
    }

    case GENES_EXPRESSION_LOADED: {
      return genesExpressionLoaded(state, action);
    }

    case GENES_EXPRESSION_ERROR: {
      return genesExpressionError(state, action);
    }

    case GENES_EXPRESSION_TYPE_UPDATE: {
      return genesExpressionTypeUpdate(state, action);
    }

    default: {
      return state;
    }
  }
};

export default genesReducer;
