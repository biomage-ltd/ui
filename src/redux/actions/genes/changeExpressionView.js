import {
  GENES_EXPRESSION_VIEW_LOADING, GENES_EXPRESSION_VIEW_LOADED, GENES_EXPRESSION_VIEW_ERROR,
} from '../../actionTypes/genes';
import loadGeneExpression from './loadGeneExpression';

import { geneOperations } from '../../../utils/geneTable/geneOperations';

const changeExpressionView = (
  experimentId, genes, componentUuid, genesOperation,
) => async (dispatch, getState) => {
  const oldData = getState().genes.expression.views[componentUuid]?.data;

  let newData = genes;

  if (genesOperation === geneOperations.ADD && oldData) {
    newData = Array.from(new Set(genes.concat(oldData)));
  }
  if (genesOperation === geneOperations.REMOVE && oldData) {
    newData = oldData.filter((gene) => !genes.includes(gene));
  }
  if (genesOperation === geneOperations.OVERWRITE && oldData) {
    newData = genes;
  }

  // Dispatch loading state
  dispatch({
    type: GENES_EXPRESSION_VIEW_LOADING,
    payload: {
      experimentId,
      genes: newData,
      componentUuid,
    },
  });

  try {
    if (genes.length > 0) {
      await dispatch(loadGeneExpression(experimentId, genes));
    }

    // Dispatch loaded state.
    dispatch({
      type: GENES_EXPRESSION_VIEW_LOADED,
      payload: {
        experimentId,
        genes,
        componentUuid,
      },
    });
  } catch (e) {
    // Dispatch error state.
    dispatch({
      type: GENES_EXPRESSION_VIEW_ERROR,
      payload: {
        experimentId,
        componentUuid,
        error: "Couldn't update the view of this component.",
      },
    });
  }
};

export default changeExpressionView;
