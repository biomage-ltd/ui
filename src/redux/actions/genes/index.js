import loadGeneExpression from './loadGeneExpression';
import changeGeneSelection from './changeGeneSelection';
import loadPaginatedGeneProperties from './loadPaginatedGeneProperties';
import loadMarkerGenes from './loadMarkerGenes';
import markerGenesLoaded from './markerGenesLoaded';
import markerGenesError from './markerGenesError';

export {
  loadGeneExpression,
  changeGeneSelection,
  loadPaginatedGeneProperties,
  loadMarkerGenes,
  markerGenesLoaded,
  markerGenesError,
};
