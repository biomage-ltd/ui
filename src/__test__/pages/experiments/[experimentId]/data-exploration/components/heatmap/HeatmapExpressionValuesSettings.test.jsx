import React from 'react';
import { configure, mount } from 'enzyme';
import { Provider } from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Radio } from 'antd';
import HeatmapExpressionValuesSettings from '../../../../../../../pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapExpressionValuesSettings';

import { GENES_EXPRESSION_TYPE_UPDATE } from '../../../../../../../redux/actionTypes/genes';

jest.mock('localforage');

const mockStore = configureMockStore([thunk]);
configure({ adapter: new Adapter() });

let component;

const geneState = {
  genes: {
    expression: {
      views: {},
      expressionType: 'raw',
    },
  },
};

describe('HeatmapExpressionValuesSettings', () => {
  afterEach(() => {
    component.unmount();
  });

  it('renders correctly', () => {
    const store = mockStore({
      ...geneState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapExpressionValuesSettings experimentId='123' />
      </Provider>,
    );

    // Should be rendered.
    expect(component.find('HeatmapExpressionValuesSettings').length).toEqual(1);

    // With two buttons.
    const buttons = component.find('Radio');
    expect(buttons.length).toEqual(2);

    // With the right keys.
    expect(buttons.at(0).props().value).toEqual('raw');
    expect(buttons.at(1).props().value).toEqual('zScore');

    // The selected one should be `raw`.
    expect(component.find(Radio.Group).props().value).toEqual('raw');
  });

  it('choosing another setting will trigger the appropriate action', async () => {
    const store = mockStore({
      ...geneState,
    });

    component = mount(
      <Provider store={store}>
        <HeatmapExpressionValuesSettings experimentId='123' />
      </Provider>,
    );

    // Should be rendered.
    expect(component.find('HeatmapExpressionValuesSettings').length).toEqual(1);

    // When a separate expression value by is selected...
    const { onChange } = component.find(Radio.Group).props();
    onChange({ target: { value: 'zScore' } });

    // The store should update.
    const action = store.getActions()[0];
    expect(action.type).toEqual(GENES_EXPRESSION_TYPE_UPDATE);
    expect(action).toMatchSnapshot();
  });
});
