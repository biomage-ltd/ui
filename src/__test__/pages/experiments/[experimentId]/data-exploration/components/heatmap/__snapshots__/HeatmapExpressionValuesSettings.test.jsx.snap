// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatmapExpressionValuesSettings choosing another setting will trigger the appropriate action 1`] = `
Object {
  "payload": Object {
    "expressionType": "zScore",
  },
  "type": "genes/expressionTypeUpdate",
}
`;
